### ISSUE TITLE
Internationalize (i18n) strings emitted by built-in converters

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#1129]
Localize (i.e., translate) strings emitted by built-in converters and document how this feature works.

We'll need two things. First, we need to make sure that any and all strings that are emitted read from an attribute. This is true in most cases, such as `note-caption`. Next, we need to decide the best way to locate and read in attributes based on language selection. While these strings can be fed in one by one into the attributes, it's rather inconvenient and does not encourage reuse. We want to eventually have support for many languages inside of Asciidoctor like AsciiDoc Python.

**Comment by mojavelinux** [comment]
> we need to make sure that any and all strings that are emitted read from an attribute.

This part is complete.

**Comment by mojavelinux** [comment]
> While these strings can be fed in one by one into the attributes, it's rather inconvenient and does not encourage reuse. 

All these attributes, and when to use them, are now documented. See

* http://asciidoctor.org/docs/user-manual/#builtin-attributes-i18n
* http://asciidoctor.org/docs/user-manual/#customizing-built-in-labels

**Comment by ciampix** [comment]
I would like to contribute suggesting translation strings for my language and doing so suggesting other people to do the same to create a common multi-language template in the form:

    ifeval::["{lang}" == "de"]
    :caution-caption: Achtung
    ...
    endif::[]

    ifeval::["{lang}" == "it"]
    :caution-caption: Attenzione
    ...
    endif::[]

This document will be useful right now and could be used afterwards for creating the "right thing" piking up and using the already translated strings.
Do you think it is a good idea?
May I start creating this document?

**Comment by mojavelinux** [comment]
Indeed! Let's put the attributes in the file `data/locale/attributes.adoc` for now. Once we add native support for loading language files, we can split the contnet into multiple files.

**Comment by ciampix** [comment]
Done PR  #1743

**Comment by jniggemann** [comment]
More or less related: Is it possible to translate the word "Chapter" in titles?
ATM I have titles that read

> Chapter 4. Beschreibung der Optionen

**Comment by mojavelinux** [comment]
@jniggemann You are probably referring to Asciidoctor PDF. Yes, it's possible. See https://github.com/asciidoctor/asciidoctor-pdf/blob/master/docs/theming-guide.adoc#theme-related-document-attributes

**Comment by jniggemann** [comment]
@mojavelinux Err, you're right, wrong project - sorry and thank you for pointing me in the right direction.

**Comment by mojavelinux** [comment]
@jniggemann Not totally wrong. We are including the chapter-label translations in this file, but they only apply to Asciidoctor PDF atm.

**Comment by hedrok** [comment]
I suggest adding option --prepend-file to asciidoctor that simply prepends content of file to each processed file as if each of them included this file in the first line. Then it will be easy to use attributes.adoc. It will be useful for other things too. What do you think?

Or have I missed such option? (I've been searching for it for some time).

I'm migrating my project from asciidoc now and don't like the idea of adding include::attributes.adoc[] to each file...

**Comment by mojavelinux** [comment]
This can be achieved using a Preprocessor. Another option is to simply concat files together before passing them to Asciidoctor (which is effectively the same as a preprocessor).

**Comment by mojavelinux** [comment]
Now that all messages go through the logger (as of 1.5.7), this is very possible. What we're still missing that could help is an error code or label that's associated with the message. We could consider doing that now.

**Comment by michael-o** [comment]
As a user I wanted to make use of https://raw.githubusercontent.com/asciidoctor/asciidoctor/master/data/locale/attributes.adoc. After reading the docs which really recommends copying that file and adding to all of my files, I opted to throw that away because this is just ridiculous. I'd be happy with something like this:

    asciidoctor -a lang=de -a toc=left --search-path /usr/local/lib/ruby/gems/2.4/gems/asciidoctor-1.5.6.2/data index.adoc

better yet

    asciidoctor -a lang=de -a toc=left -a lang-attributes index.adoc

**Comment by mojavelinux** [comment]
You could write a preprocessor extension to load the file from the install path of the gem and prepend it to the source. You can get the idea from how we load the man page. https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/cli/options.rb#L143-L170

**Comment by michael-o** [comment]
@mojavelinux Thanks for the pointer, shouldn't something be included by default? Akin to LaTeX's include directories. Given that my Ruby knowledge is not existing, I opted for the old-fashioned way:

    tmp_dir=$(mktemp -d) ; echo "include::\(find /usr/local/lib/ruby/gems/2.4/gems/ -name attributes.adoc)[]" | cat - index.adoc > $tmp_dir/index.adoc ; cd $tmp_dir ; asciidoctor ...

**Comment by mojavelinux** [comment]
@michael-o yes, that's why this issue is still open. The remaining task is to load the locale attributes file from the gem based on the value of the lang attribute.

(The logic for this would be similar to what we're doing to load the manpage).
